packages/note-access/src/metadata/toString.js

```diff
@@ -0,0 +1,39 @@
+import { toChecksumAddress } from 'web3-utils';
+import { DYNAMIC_VAR_CONFIG_LENGTH, MIN_BYTES_VAR_LENGTH } from '../config/constants';
+import config from '../config/metadata';
+
+const toConfigVar = (num) => num.toString(16).padStart(DYNAMIC_VAR_CONFIG_LENGTH, '0');
+const ensureMinVarSize = (str, len) => str.padStart(Math.max(len || str.length, MIN_BYTES_VAR_LENGTH), '0');
+
+export default function toString(metadataObj) {
+    const metadata = {};
+    let accumOffset = 0;
+    config.forEach(({ name, length, startAt }) => {
+        let data = '';
+        if (metadataObj[name] || (!startAt && length)) {
+            const dataArr = Array.isArray(metadataObj[name]) ? metadataObj[name] : [metadataObj[name] || ''];
+            data = dataArr
+                .map((v) => (name === 'addresses' && toChecksumAddress(v)) || v)
+                .map((v) => `${v}`.replace(/^0x/, ''))
+                .map((v) => ensureMinVarSize(v, length))
+                .join('');
+        }
+
+        if (startAt) {
+            // a character is half a bytes
+            // the total length should always be even
+            metadata[startAt] = toConfigVar(accumOffset / 2);
+
+            const numberOfVars =
+                length !== undefined ? data.length / Math.max(length, MIN_BYTES_VAR_LENGTH) : Math.min(1, data.length);
+            data = `${toConfigVar(numberOfVars)}${data}`;
+        }
+
+        metadata[name] = data;
+        accumOffset += startAt ? data.length : length;
+    });
+
+    const str = config.map(({ name }) => metadata[name]).join('');
+
+    return `0x${str}`;
+}
```

Comment:
I had a really hard time getting to grips with what was happening here, and it struck me as quite brittle to any change in ABI. I think giving every type a `_toString` method (see `config/metadata.js` comments) isolates any particularities with the types over there, and makes this function much simpler: 

```
import { DYNAMIC_VAR_CONFIG_LENGTH } from '../config/constants';
import config from '../config/metadata';

const to32ByteOffset = (num) => num.toString(16).padStart(DYNAMIC_VAR_CONFIG_LENGTH, '0');

export default function toString(metadata) {
    let totalDataLength = 0;
    const stringifiedMetadata = config.reduce((acc, type) => {
        const { name, length, startAt: offset } = type;
        acc[name] = type._toString(metadata[name]);

        // Need to find and set the offset of dynamic types.
        if (offset) {
            // A character is half a byte
            acc[offset] = to32ByteOffset(totalDataLength / 2);
        }

        totalDataLength += (offset) ? acc[name].length : length;
        return acc;
    }, {});

    const str = config.map(({ name }) => stringifiedMetadata[name]).join('');

    return `0x${str}`;
}
```

Reply:
Agreed, have implemented an `encodeMetaDataToString()` function which tries to make the code more accessible